### ISSUE TITLE
(Pre)Proposal: wasi-math

### ISSUE THREAD

**Issue by oovm** [WebAssembly/WASI#587]
I would like to make some comments here: Is it necessary to develop [wasi-math](https://github.com/oovm/wasi-math)?

---

## Motivation

Many people want to use some slightly complex mathematical functions in wasm, but their high-performance implementation is actually not simple, especially in wasm.

- https://github.com/WebAssembly/design/issues/1503
- https://github.com/WebAssembly/wasi-libc/issues/92

So I'm wondering if it's possible to add math functions to the next stage of wasi (in preview3)?

In this case, the host platform can directly use hardware acceleration on platforms that contain corresponding instructions, and use programming implementation on platforms that do not support it.

---

## Goals(MVP)

The initial goal is to achieve the same computing power as [Javascript Math](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math).

Including abilities such as large integers, trigonometric functions, inverse trigonometric functions, exponential logarithms, etc.

## Non-Goals

- Already included in WASM
   - min, max
- The algorithm is extremely simple
   - Fibonacci Sequence, 3n+1
- Difficult to implement the optimal algorithm
    - Prime number, determine whether it is a prime number

## To be discussed

- The number theory function is a very commonly used function in cryptography. 
  - I don’t know what component it should belong to.
  - Mathematics? Cryptography? Or a dedicated number theory proposal?
- Vector operations, quaternions
   - is very common in games, and I focused on the needs of platforms such as Unity.

**Comment by spotandjake** [comment]
I wonder about the overhead from the component model here, when discussing regular system interfaces the overhead of the system call is likely to be far greater than any overhead from the shared nothing nature of the component model, but when talking about fast math operations that overhead is a lot more relevant, we are working with numbers though which one would assume in most languages would be on the stack, and passed by value using standard encodings but I think the discussion is still important. When working on [grains number library](https://grain-lang.org/docs/stdlib/number) I found trig functions and the power operation (especially in relation to floats) were the hardest things with all other functions being trivial so I think these should certainly be considered. I do wonder if though if this is a scenario best left to someone just building a wit-interface and composing components with wac as it's not system related and what you want really depends on your use case.

**Comment by oovm** [comment]
My consideration is mainly two application scenarios

1. The first is the case of a host function, in which the function is passed in and called by the host
The libc or standard library linked by the host already has a basic math library, which does not cause too much extra work
2. The second is the case of a guest function, where the host may be resource-constrained and unable to provide such an interface, or only provides an old wasm runtime
Then in the distant future, the optimizer can link to the implementation of a polyfill library and perform [cross-component link time optimization](https://github.com/WebAssembly/binaryen/issues/6728#issuecomment-2225913156) to inline them

No matter which solution is adopted, it makes sense to write a set of interfaces with wit first.

I defined some of [my own wit libraries](https://github.com/oovm/wasi-math) here, all functions are divided into f32 and f64 versions.

**Comment by oovm** [comment]
A potentially controversial point is whether to include bigint. My main considerations are:

1. It already exists in js, and no interface implementation will be added
2. It is dependent on cryptography
3. Handwritten wasm implementation may be difficult to implement FFT, NTT hardware acceleration

The less certain point is whether there will be a `wasm:js-bigint` proposal similar to `wasm:js-string`, in which case there is a conflict.

**Comment by programmerjake** [comment]
> A potentially controversial point is whether to include bigint. My main considerations are:

> 2. It is dependent on cryptography

assuming you meant that cryptography is dependent on bigint. bigint implementations for cryptography are quite different than bigint implementations for general arithmetic, since they have to use different often-slower algorithms to prevent timing side-channels (leaking the secret information your processing by being able to observe how long it takes over the network).

**Comment by spotandjake** [comment]
As component support gets better and wit interfaces become more ubiquitous I don't see the benefit of having this as a wasi library as opposed to just composing your application with a component for doing math.

**Comment by ricochet** [comment]
Thank you for proposing this. WASI proposals need a champion to advance through the phase process, and this doesn't currently have one. If you or anyone else wants to champion it, then the next appropriate step is to follow [phase 0 pre-proposal process](https://github.com/WebAssembly/WASI/blob/main/CONTRIBUTING.md#0-pre-proposal-individual-contributor).

In the meantime, this doesn't have to wait on WASI. WASI is built on the Wasm [Component Model](https://github.com/WebAssembly/component-model), whose worlds and interfaces are designed to be modular. An interface can be described, implemented, and shipped independently of WASI itself. Building and iterating on your own is valuable. It's often the fastest way to learn whether the design holds up, and is a great way to build community support behind a given interface or design before pursuing standardization.
